### ISSUE TITLE
Alternator: support Projection in GSI and LSI

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#5036]
When creating a GSI or LSI with `CreateTable`, the DynamoDB API provides a `Projection` parameter determining what gets copied to the materialized view: Just the keys, all the attributes, or the keys plus a few select attributes. We should support this parameter - currently we ignore it and always project all attributes.

Note that for LSI, there is a delicate issue: Even when only specific attributes are projected, DynamoDB allows to read also unprojected attributes (we verify this in `test_lsi_get_not_projected_attribute`). We can implement this either by projecting everything (as we do today) and not returning the unprojected columns when not specifically requested, or by doing a second query to get the unprojected columns. The second approach is probably closer to what DynamoDB does. Note that the second query is always on the same shard as the first one (same partition key).

We have existing xfailing tests for this issue:
1.  In `test_gsi.py`:  `test_gsi_projection_keys_only`, `test_gsi_projection_include`, `test_gsi_missing_projection_type`, 
2. In `test_lsi.py`: `test_lsi_get_all_projected_attributes`.

**Comment by tzach** [comment]
> or by doing a second query to get the unprojected columns.

AFAIU, this is what DynamoDB is doing
" However, to retrieve any additional attributes, DynamoDB must perform additional read operations against the Thread table. These extra reads are known as fetches, and they can increase the total amount of provisioned throughput required for a query."
https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/LSI.html#LSI.Projections

**Comment by nyh** [comment]
It will be significantly easier to implement KEYS_ONLY projection, because it just means not to include the ":attrs" in the view - no need to slice it. So we should start with implementing the KEYS_ONLY option.

**Comment by doublex** [comment]
@nyh 
As a temporary workaround:
Is it possible to use an "ALTER MATERIALIZED VIEW" statement to simulate `KEYS_ONLY`?

**Comment by nyh** [comment]
> @nyh As a temporary workaround: Is it possible to use an "ALTER MATERIALIZED VIEW" statement to simulate `KEYS_ONLY`?

CQL's `ALTER MATERIALIZED VIEW` cannot change the SELECTion of the view, so you cannot use it to get rid of the `:attrs` column in the view (which is what you'd want to do). 
One hack that may still be possible without any code changes is to use CQL to `DROP MATERIALIZED VIEW` the view created by Alternator, and create a new one with `CREATE MATERIALIZED VIEW` whose SELECT doesn't include `:attrs`. This may work, but I didn't test it actually does. One of the questions is whether the code that _reads_ from a view is ok with the fact there is no `:attrs` column in that table.

**Comment by doublex** [comment]
Thanks! Yes, it is about removing `:attrs`.
Droping and creating a new materialized-view is perfectly fine.
Is there a way to inspect the old materialized-view-select-statement created by Alternator?
I would like to test and report back.

**Comment by doublex** [comment]
@nyh 
Thank you for your assistance.
ScyllaDB does not allow to create the alternator materialized-view manually.
Alternator could create the appropriate materialized view.

E.g.:
````
base-table "data":
{
    primary_key: binary,
    si_hash_key: binary,
    si_range_key: binary,
    payload: binary,
}
CREATE MATERIALIZED VIEW alternator_data."data:index" AS
    SELECT primary_key, si_hash_key, si_range_key
    FROM alternator_data.data
    WHERE si_hash_key IS NOT NULL AND si_range_key IS NOT NULL
    PRIMARY KEY (si_hash_key, si_range_key, primary_key)
    WITH CLUSTERING ORDER BY (si_range_key ASC, primary_key ASC);
````

Error message, docs: https://opensource.docs.scylladb.com/stable/cql/mv.html
`
InvalidRequest: Error from server: code=2200 [Invalid query] message="Cannot include more than one non-primary key column 'si_range_key' in materialized view primary key"
`

The `create materialzed view`statement was taken from `DESCRIBE MATERIALIZED VIEW alternator_data."data:index";`

**Comment by mykaul** [comment]
> It will be significantly easier to implement KEYS_ONLY projection, because it just means not to include the ":attrs" in the view - no need to slice it. So we should start with implementing the KEYS_ONLY option.

Agreed, looks like an easy win with high 'profit' - but I don't know how common is the use of KEYS_ONLY.

**Comment by pauladeola9** [comment]
@nyh 
What's the status of this work?

**Comment by nyh** [comment]
> [@nyh](https://github.com/nyh) What's the status of this work?

In general, the purpose of the bug-tracker issue is that if someone is working on a feature, the issue is updated. The fact it wasn't is an indication that there is no status to report. This feature is still not supported, not even the `KEYS_ONLY` case. It is not a hard feature to implement, but we just didn't get around to it.

**Comment by nyh** [comment]
Draft pull request implementing `ProjectionType=KEYS_ONLY` (but not `ProjectionType=INCLUDE`): https://github.com/scylladb/scylladb/pull/29601.

**Comment by doublex** [comment]
Great news!
